### PR TITLE
Prevent stale adapter:collectionView corruptions

### PR DESCRIPTION
**Pull request by rnystrom**
## Changes in this pull request

This becomes an issue pretty easily in embedded `IGListAdapter` environments (lists in lists). IGListKit creates a single `IGListSectionController` instance per object, but cells are reused within `IGListCollectionView`. If you assign `adapter.collectionView = cell.collectionView` (like we recommend in our [example](https://github.com/Instagram/IGListKit/blob/master/Examples/Examples-iOS/IGListKitExamples/SectionControllers/HorizontalSectionController.swift#L40)), then you can have many adapters with a weak ref to the **same collection view**.

Obviously when updates happen across all embedded lists, adapters could potentially update the same collection view and apply a corrupted batch update.

I proposed using `<objc/runtime.h>` and using associated objects, but since `OBJC_ASSOCIATION_ASSIGN` is not zerod-on-release reference (its just `unsafe_unretained`) I'd have to make a weak wrapped object, which is overkill.

Instead I think a global weak:weak map is fine. We don't have to worry about locking b/c of the main-thread affinity of the framework. Please push back on me if you have a better idea.

Issue fixed: #503

### Checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I added tests, an experiment, or detailed why my change isn't tested.
- [x] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.

### CODE REVIEW

**Comment by rnystrom** [comment]
Importing b/c our CI is faster than Travis.

**Comment by jessesquires** [comment]
> our CI is faster than Travis.

lol

File: Source/IGListAdapter.m [hunk 1]
```diff
@@ -62,13 +62,23 @@ - (IGListCollectionView *)collectionView {
     return (IGListCollectionView *)_collectionView;
 }
 
+static void *kIGListCollectionViewAdapterKey = &kIGListCollectionViewAdapterKey;
```
Comment (jessesquires) [inline Source/IGListAdapter.m]: where is this used? 🤔

File: Source/IGListAdapter.m [hunk 2]
```diff
@@ -62,13 +62,23 @@ - (IGListCollectionView *)collectionView {
...
+
 - (void)setCollectionView:(IGListCollectionView *)collectionView {
     IGAssertMainThread();
-    IGParameterAssert([collectionView isKindOfClass:[IGListCollectionView class]]);
```
Comment (jessesquires) [inline Source/IGListAdapter.m]: unrelated to this PR? 🤔

File: Source/IGListAdapter.m [hunk 3]
```diff
@@ -62,13 +62,23 @@ - (IGListCollectionView *)collectionView {
...
 
     // if collection view has been used by a different list adapter, treat it as if we were using a new collection view
     // this happens when embedding a IGListCollectionView inside a UICollectionViewCell that is reused
     if (_collectionView != collectionView || _collectionView.dataSource != self) {
+        // if the collection view was being used with another IGListAdapter (e.g. cell reuse)
+        // destroy the previous association so the old adapter doesn't update the wrong collection view
+        static NSMapTable<IGListCollectionView *, IGListAdapter *> *globalCollectionViewAdapterMap = nil;
```
Comment (jessesquires) [inline Source/IGListAdapter.m]: we're main-thread-only, but should we still consider using `dispatch_once` ?

**Comment by jessesquires** [comment]
💯

File: Source/IGListAdapter.m (hunk 1 above)
Comment (rnystrom) [inline Source/IGListAdapter.m]: It's not 😅

File: Source/IGListAdapter.m (hunk 2 above)
Comment (rnystrom) [inline Source/IGListAdapter.m]: Have to remove so we can set to `nil`

File: Source/IGListAdapter.m (hunk 3 above)
Comment (rnystrom) [inline Source/IGListAdapter.m]: Totally fine if main-thread affined, `dispatch_once` will add unused overhead

**Comment by heumn** [comment]
thisisgiphy-reaction-audience-l4HodBpDmoMA5p9bG

**Comment by ninokierulf** [comment]
after applying patch f639cdd,
the embedded collectionview's cell's focused/selected state is being reset.

**Pull request closed without merging**